Use vim 8 timers for continuous checking of debugger output

IngoMeyer441 opened the pull request:
Fixes #32 
Native vim timers are used if availabe. Otherwise the old code is used. Vim timers avoid the problems mentioned in #32 .

Commit:
Use vim 8 timers for continuous checking of debugger output

idanarye commented on autoload/vebugger.vim:
```diff
@@ -268,26 +270,39 @@ function! vebugger#createDebugger(command)
-        autocmd CursorHold * call s:debugger.invokeReading()
-    augroup END
+    if has('timers')
+        let s:timer_id = timer_start(500, function('ReadingTimerCallback', [s:debugger]), {'repeat': -1})
```
Neovim does not support partial functions. Since `s:debugger` is global, just make the callback use it.

idanarye commented on autoload/vebugger.vim:
```diff
@@ -268,26 +270,39 @@ function! vebugger#createDebugger(command)
 endfunction
 
+if has('timers')
+    function! ReadingTimerCallback(debugger, timer_id)
```
Rename this to `s:readingTimerCallback`

IngoMeyer441 replied:
The timer callback part seems to be a bit tricky. I had some problems to get it running at all, since `start_timer` does not accept script local functions directly (it gives `E120: Using <SID> not in a script context`). But you are right, it is not necessary to have both a local function and a partial. After some tests I found out: 

You can use either use

- a global function and pass it to `start_timer` or
- a script local function and pass a partial to `start_timer`. You are right, there is no need to bind the `s:debugger` variable.

First I thought that timer callbacks mustn't access any script local attributes; so I coded that ugly solution with a global function and a partial.

Isn't it possible at all to use partials (or anything similiar) in neovim? It would be quite ugly to pick the solution with a global function.

idanarye replied:
Are you sure about `start_timer` not accepting script local functions? Because I just checked, and this works fine in both Vim8 and Neovim:

```vim
let s:x = 0

function! s:foo(timerId)
	let s:x += 1
	echo 'foo' s:x a:timerId
endfunction

call timer_start(1000, function('s:foo'), {'repeat': 3})
```

[Neovim does not support partials yet](https://github.com/neovim/neovim/issues/5065). The only other way to pass variables together with a function is an object - but `start_timer` does not support that. Neovim will probably get partials in the near future, but I'm not willing to drop backward compatibility yet([I even gave up some encapsulation to support Vim7.3](https://github.com/idanarye/vim-vebugger/issues/5)).

At some point in the future I want to bump the major version and rewrite the message bus to use jobs. This will make the timer redundant. Until then - we'll have to use a global variable.

IngoMeyer441 replied:
Ah OK, I thought that `function` in an expression would always create a partial - in your example a partial that does not bind any argument. In my tests, I used `start_timer` with a string of the function name directly:
```vim
timerId = start_timer(500, 's:foo')
```
That syntax only works with global functions. Sorry, I'm not very familiar with vimscript but I'm getting better I think. :) 

Since your example works in neovim, everything is fine and I will update my patch tomorrow.

idanarye replied:
Yea, Vimscript is weird like that... I would actually expect `s:foo` will be the function pointer - or at the very least `s:.foo`...

idanarye commented on autoload/vebugger.vim:
```diff
@@ -268,26 +270,39 @@ function! vebugger#createDebugger(command)
-    augroup vebugger_shell
-        autocmd!
-    augroup END
+    if has('timers') && exists('s:timer_id')
```
Use two `if` statements. If this is called before the first debugger was created, it'll run the autogroup cleanup code. Not that it'll damage anything - but if you are going to do things you should do them right.

IngoMeyer441 replied:
You are completely right. It was a bit late tonight... ;)

idanarye commented on autoload/vebugger.vim:
```diff
@@ -268,26 +270,39 @@ function! vebugger#createDebugger(command)
-        autocmd CursorHold * call s:debugger.invokeReading()
-    augroup END
+    if has('timers')
+        let s:timer_id = timer_start(500, function('ReadingTimerCallback', [s:debugger]), {'repeat': -1})
```
Also - please rename it to `s:timerId`. Naming convention in the same project should be consistent...

IngoMeyer441 replied:
Already done :) The next version of the pull request will have `timerId`.

IngoMeyer441 commented:
Is there still something that should be improved?

IngoMeyer441 commented:
Ok I noticed an issue. Evaluated expressions are no longer printed at the window bottom.
I think the reason for the problem is that vim no longer waits till a key is pressed when an evaluated expression is printed. Vim no longer waits since `printEvaluatedExpression` is now running asynchronously.
If the trailing `\n` in the echo message is removed (https://github.com/idanarye/vim-vebugger/blob/develop/autoload/vebugger/std.vim#L239) the output works fine. Is that `\n` really necessary?

idanarye commented:
Sorry, I did not notice you updated the PR...

The purpose of the `\n` is to pause Vim so that the user could see the message. Not really necessary - especially if it prevents them from seeing the message at all. I am a bit worried though about multiline messages...

BTW, the `let g:echo=...` line was for debugging purposes and I forgot to remove it - so don't bother to update it to match your changes to the actual `echo`. You can remove it if you want.

IngoMeyer441 commented:
Hmm, I could just remove the trailing `\n` and oneline output will work fine. But I don't know how to fix that for multiline output. I think for multiline output only a new window that is reserved for eval output can guarantee that the user can see the whole message.

idanarye commented:
OK, I found a way:
```vim
                " The text to echo:
                let l:echo = l:evaluatedExpression.expression.': '.l:evaluatedExpression.value
                let l:echoLines = split(l:echo, '\r\n\|\n\|\r')
                if len(l:echoLines) == 1
                    call add(l:echoLines, '')
                endif

                if has('timers')
                    " Convert to an expression we can use with feedkeys.
                    " string(l:echo) won't do because it uses single-quotes which
                    " do not escape newlines
                    let l:echoJoinExpr = 'join('.string(l:echoLines).', "\n")'

                    let l:echoKeys = ':echo '.l:echoJoinExpr."\<Cr>"
                    if mode() == 'n'
                        " Call echo normaly
                        call feedkeys(l:echoKeys)
                    elseif mode() == 'i'
                        " Execute command in insert mode
                        call feedkeys("\<C-o>".l:echoKeys)
                    endif
                    " NOTE: Other modes are not supported
                else
                    " Without timer support, feedkeys won't work and we have
                    " to echo directly
                    echo join(l:echoLines, "\n")
                endif
```

It's a bit complex, and only works with normal and insert mode - so if the user quickly moves to a different mode it won't work. But this is rarely a problem.

IngoMeyer441 commented:
Wow, great idea to use `feedkeys` as a kind of synchronization mechanism. :)
I tried to apply your code and got stuck at line [233](https://github.com/idanarye/vim-vebugger/blob/develop/autoload/vebugger/std.vim#L233) in `std.vim`:
```vim
if empty(get(l:evaluatedExpression,'expression'))
    echo l:evaluatedExpression.value."\n"
```
Is it possible that `expression` does not exist? Or is it just for the case a debugger handler was implemented faulty?

Commit:
Use feedkeys for echoing evaluated expressions

Echo commands in vim 8 timer handlers do not wait for the user to
confirm. Echoing via feedkeys circumvents this problem.

IngoMeyer441 commented:
I've adopted your solution and tested it -> works great!
Unfortunately, some debugger frontends don't process multiline evaluation output. For example, the PDB frontend only reads the first line of a multiline numpy array output.

idanarye commented:
`expression` could not exist if a debugger was printing the evaluated values without the expression to be evaluated. I can't remember which debugger was using it though...

As for debuggers that don't process multiline evaluation output - sadly there is little we can do about it. I'm not going to start parsing Python literals for delimiters!
